src/librustc_codegen_ssa/back/link.rs

```diff
@@ -70,35 +71,29 @@ pub fn link_binary<'a, B: ArchiveBuilder<'a>>(
             }
         });
 
-        let tmpdir = TempFileBuilder::new()
-            .prefix("rustc")
-            .tempdir()
-            .unwrap_or_else(|err| sess.fatal(&format!("couldn't create a temp dir: {}", err)));
-
         if outputs.outputs.should_codegen() {
+            let tmpdir = TempFileBuilder::new()
+                .prefix("rustc")
+                .tempdir()
+                .unwrap_or_else(|err| sess.fatal(&format!("couldn't create a temp dir: {}", err)));
+            let path = TempDir::new(tmpdir, sess.opts.cg.save_temps);
```

Comment:
Can we use a different name for this? `MaybeTempDir` or something? The boolean argument to `new` is surprising if you're used to `tempfile::TempDir`.